wp/lib/bap_wp/src/precondition.ml

```diff
@@ -155,12 +155,137 @@ let bv_to_bool (bv : Constr.z3_expr) (ctx : Z3.context) (width : int) : Constr.z
+      Env.new_z3_expr env ~name:("unknown_" ^ str) typ, env
+    | Ite (cond, yes, no) ->
+      debug "Visiting ite: if %s\nthen %s\nelse %s%!"
+        (Exp.to_string cond) (Exp.to_string yes) (Exp.to_string no);
+      let cond_val, env = exp_to_z3_body cond env in
+      let cond_size = BV.get_size (Expr.get_sort cond_val) in
+      let yes_val, env = exp_to_z3_body yes env in
+      let no_val, env = exp_to_z3_body no env in
+      Bool.mk_ite ctx (bv_to_bool cond_val ctx cond_size) yes_val no_val, env
+    | Extract (high, low, exp) ->
+      debug "Visiting extract: High:%d Low:%d Exp:%s%!" high low (Exp.to_string exp);
+      let exp_val, env = exp_to_z3_body exp env in
+      BV.mk_extract ctx high low exp_val, env
+    | Concat (w1, w2) ->
+      debug "Visiting concat: %s ^ %s%!" (Exp.to_string w1) (Exp.to_string w2);
+      let w1_val, env = exp_to_z3_body w1 env in
+      let w2_val, env = exp_to_z3_body w2 env in
+      BV.mk_concat ctx w1_val w2_val, env
+  in
+  let a, v = Env.mk_exp_conds env exp in
+  let assume = List.map a ~f:Constr.mk_constr in
+  let verify = List.map v ~f:Constr.mk_constr in
+  let z3_exp, new_env = exp_to_z3_body exp env in
+  z3_exp, assume, verify, new_env
+
 let set_fun_called (post : Constr.t) (env : Env.t) (tid : Tid.t) : Constr.t =
   let ctx = Env.get_context env in
   let fun_name = Bool.mk_const_s ctx (Env.get_called env tid |>
                                       Option.value_exn ?here:None ?error:None ?message:None) in
   Constr.substitute_one post fun_name (Bool.mk_true ctx)
 
+let get_stack_ptr_offsets (sub : Sub.t) : Exp.t list =
+  let blks = Term.to_sequence blk_t sub in
+  let ret_block =
+    Seq.find blks ~f:(fun b ->
+        let jmps = Term.to_sequence jmp_t b in
+        Seq.exists jmps ~f:(fun j ->
+            match Jmp.kind j with
+            | Ret _ -> true
+            | _ -> false))
+  in
+  match ret_block with
+  | None ->
+    warning "Sub %s has no return" (Sub.name sub);
+    []
+  | Some blk ->
+    (* FIXME: We are assuming that the stack pointer is RSP here. *)
+    let rsp = Var.create "RSP" reg64_t in
+    let defs = Term.to_sequence def_t blk in
+    Seq.fold defs ~init:[] ~f:(fun offsets def ->
+        (* Looks for instructions in the form RSP := RSP +/- offset *)
+        if Var.equal (Def.lhs def) rsp then begin
+          let rhs = Def.rhs def in
+          match rhs with
+          | BinOp (PLUS, Var v, Int _) ->
+            if Var.equal v rsp then
+              rhs :: offsets
+            else offsets
+          | BinOp (MINUS, Var v, Bil.Types.Int w) ->
+            if Var.equal v rsp then begin
+              warning "RSP is being decremented by %s" (Word.to_string w) ;
+              rhs :: offsets
+            end else offsets
+          | _ -> offsets
+        end else
+          offsets)
+
+let increment_stack_ptr (post : Constr.t) (env : Env.t) (offsets : Exp.t list)
+  : Constr.t =
```

Comment:
Probably not a big deal, but you could pass the `Env.t` here, instead of throwing it away. Here the `Env.t` doesn't get updtated, so it doesn't change, but it would be more uniform.